aten/src/ATen/core/CachingHostAllocator.h

```diff
@@ -246,13 +249,12 @@ struct CachingHostAllocatorImpl {
   }
 
   virtual B* get_free_block(size_t size) {
-    std::lock_guard<std::mutex> g(free_list_mutex_);
-    B key(size);
-    auto it = free_list_.lower_bound(&key);
-    if (it != free_list_.end()) {
-      B* block = *it;
+    auto index = size_index(size);
```

Comment:
If I'm not mistaken, this indeed changes the logic of finding the suitable free block. Previously, the entire free list was searched for an appropriate lock; for example, if 64 bytes were requested but only 128 bytes were available in the free list, the 128 bytes block would be reused. However, in the new implementation, a new 64 bytes allocation would be performed instead of reusing the 128 bytes block.
The new implementation seems more reasonable since the host caching allocator doesn't split the excess blocks to improve storage utilization.
If you could share the data on the performance or memory utilization improvements, that would be great. Others look good to me.

Reply:
Based on some initial testing, we didn't see much difference in the memory, it was mostly in the noise margin.